server/src/main/java/org/elasticsearch/common/lucene/search/function/ScriptScoreFunction.java

```diff
@@ -58,27 +58,26 @@ public DocIdSetIterator iterator() {
 
     private final Script sScript;
 
-    private final SearchScript.LeafFactory script;
+    private final ScoreScript.LeafFactory script;
 
 
-    public ScriptScoreFunction(Script sScript, SearchScript.LeafFactory script) {
+    public ScriptScoreFunction(Script sScript, ScoreScript.LeafFactory script) {
         super(CombineFunction.REPLACE);
         this.sScript = sScript;
         this.script = script;
     }
 
     @Override
     public LeafScoreFunction getLeafScoreFunction(LeafReaderContext ctx) throws IOException {
-        final SearchScript leafScript = script.newInstance(ctx);
+        final ScoreScript leafScript = script.newInstance(ctx);
         final CannedScorer scorer = new CannedScorer();
-        leafScript.setScorer(scorer);
```

Comment:
I think we want to keep the scorer. Having the script access score through `get_score()` (which maps to `_score`) allows score to not be computed when it is not used. needs_score only gets us so far (whether it is used at all). With get_score, we could (although I'm not sure it is actually implemented this way yet) call the method lazily to bind _score.